cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,440 @@
+                                  B.data(),
+                                  B.shape().first,  // rows of B
+                                  &beta,            // + beta * C
+                                  C.data(),
+                                  C.shape().first,  // rows of C
+                                  num_batches));
+  return C;
+}
+
+//! A utility method to implement pointwise operations between elements of two batched matrices.
+//! @param A First matrix
+//! @param B Second matrix
+//! @param binary_op The binary operation used on elements of `A` and `B`
+//! @return The result of `A` `binary_op` `B`
+template <typename F>
+BatchedMatrix b_aA_op_B(const BatchedMatrix& A, const BatchedMatrix& B,
+                        F binary_op) {
+  if (A.shape().first != B.shape().first &&
+      A.shape().second != B.shape().second) {
+    throw std::runtime_error(
+      "Batched Matrix Addition ERROR: Matrices must be same size");
+  }
+  if (A.batches() != B.batches()) {
+    throw std::runtime_error("A & B must have same number of batches");
+  }
+
+  auto num_batches = A.batches();
+  int m = A.shape().first;
+  int n = A.shape().second;
+
+  BatchedMatrix C(m, n, num_batches, A.cublasHandle(), A.allocator(),
+                  A.stream());
+
+  LinAlg::binaryOp(C[0], A[0], B[0], m * n * num_batches, binary_op, 0);
+
+  return C;
+}
+
+//! Multiplies each matrix in a batch-A with it's batch-B counterpart.
+//! A = [A1,A2,A3], B=[B1,B2,B3]
+//! return [A1*B1, A2*B2, A3*B3]
+//! @param A Matrix `A`
+//! @param B Matrix `B`
+//! @returns The result of the batched matrix-matrix multiplication of `A` x `B`
+BatchedMatrix operator*(const BatchedMatrix& A, const BatchedMatrix& B) {
+  return b_gemm(A, B);
+}
+
+//! Adds two batched matrices together element-wise.
+//! @param Matrix A
+//! @param Matrix B
+//! @return A+B
+BatchedMatrix operator+(const BatchedMatrix& A, const BatchedMatrix& B) {
+  return b_aA_op_B(A, B, [] __device__(double a, double b) { return a + b; });
+}
+
+//! Subtract two batched matrices together element-wise.
+//! @param Matrix A
+//! @param Matrix B
+//! @return A-B
+BatchedMatrix operator-(const BatchedMatrix& A, const BatchedMatrix& B) {
+  return b_aA_op_B(A, B, [] __device__(double a, double b) { return a - b; });
+}
+
+//! Solve Ax = b for given batched matrix A and batched vector b
+//! @param Matrix A
+//! @param vector b
+//! @return A\b
+BatchedMatrix b_solve(const BatchedMatrix& A, const BatchedMatrix& b) {
+  auto num_batches = A.batches();
+  const auto& handle = A.cublasHandle();
+
+  int n = A.shape().first;
+  auto allocator = A.allocator();
+  int* P = (int*)allocator->allocate(sizeof(int) * n * num_batches, 0);
+  int* info = (int*)allocator->allocate(sizeof(int) * num_batches, 0);
+
+  BatchedMatrix Ainv(n, n, num_batches, A.cublasHandle(), A.allocator(),
+                     A.stream());
+
+  CUBLAS_CHECK(
+    cublasDgetrfBatched(handle, n, A.data(), n, P, info, num_batches));
```

Comment:
just like `cublasgemmBatched` wrapper inside `cublas_wrappers.h`, please create a similar wrapper for this one too.

Reply:
Done.